### ISSUE TITLE
Streaming without SurfaceView

### ISSUE THREAD

**Issue by pablomd314** [fyhertz/libstreaming#126]
How can I stream without using a SurfaceView? I know MediaCodec method 1 is buffer to buffer but I still get errors when not defining a SurfaceView. How can I get rid of the SurfaceView and still have the project work?

**Comment by Locnath** [comment]
It should be possible. Have you tried commenting out SurfaceView related functions from VideoStream.java?
For eg:
createCamera() would throw an exception is surface is null. 

What kind of errors do you get when not defining a surfaceView?

**Comment by Angie-Ie** [comment]
how to use media projection api with libstreaming? i try comment all camera input in lib but it get so many error.

**Comment by karlraab** [comment]
Are there any news how to achieve this? Urgently need to stream without showing the preview.

**Comment by gatosempai** [comment]
I did it, you just need to create a fake preview, example     


        SurfaceView sv;
        WindowManager wm = (WindowManager) getSystemService(Context.WINDOW_SERVICE);
        LayoutInflater inflater =(LayoutInflater)getSystemService(LAYOUT_INFLATER_SERVICE);
        RelativeLayout layout=(RelativeLayout) inflater.inflate(R.layout.fake_preview,null);
        WindowManager.LayoutParams params = new WindowManager.LayoutParams(1, 1,
                    WindowManager.LayoutParams.TYPE_SYSTEM_OVERLAY,
                    WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH,
                    PixelFormat.TRANSLUCENT);

        wm.addView(layout, params);

        sv = (SurfaceView) layout.findViewById(R.id.surfaceView_fake);
        SurfaceHolder sh = sv.getHolder();
        sv.setZOrderOnTop(true);
        sh.setFormat(PixelFormat.TRANSPARENT);

        // Configures the SessionBuilder
        mSession = SessionBuilder.getInstance()
                .setContext(getApplicationContext())
                .setAudioEncoder(SessionBuilder.AUDIO_AAC)
                .setAudioQuality(new AudioQuality(8000, 16000))
                .setVideoEncoder(SessionBuilder.VIDEO_H264)
                .setSurfaceView(sv)
                .setPreviewOrientation(90)
                .setCallback(this)
                .build();`

This code was implemented in a service without an activity

In your manifest add

`<uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW" />`

and finally give overlay permissions

`    
         if (Build.VERSION.SDK_INT >= 23) {
            if (!Settings.canDrawOverlays(this)) {
                Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                        Uri.parse("package:" + getPackageName()));
                startActivityForResult(intent, 1234);
            } 
        } else {
            //startService(new Intent(this, CameraService.class));
        }`

**Comment by azjkjensen** [comment]
@gatosempai I tried to implement your solution but I'm getting an error from the camera saying `app passed NULL surface`. Do you have any suggestions for how to fix this?

**Comment by karlraab** [comment]
It's really a long time ago, but as I remember I finally just took the solution to just put the SurfaveView into the background and put everything else in another SurfaceView on top of it. That worked for my purposes. But I needed it only for an art-installation once and not a "real" app. But maybe this information is useful for you too.

**Comment by azjkjensen** [comment]
Yeah I ended up getting it to work. Thanks for the help. It turns out that there's still an issue with no frames being sent when the screen isn't changing, but I'm working through it.

**Comment by Hansbald** [comment]
Did you ever manage to fix this?
I can take single pictures without preview, I even found a way to record a video without preview using the MediaRecorder but then I don't know how to stream that over the network

**Comment by cdemel** [comment]
As of 8.1, I have found it's possible to put the SurfaceView into a RelativeLayout, then put a same-sized opaque ImageView over that in the same RelativeLayout.
